### PR TITLE
LL-636 feat: Add tracking each org's disk usage

### PR DESCRIPTION
**Pull request by h-kanazawa**
Closes https://github.com/LearningLocker/enterprise/issues/893
Closes https://github.com/LearningLocker/enterprise/issues/894
https://ht2labs.myjetbrains.com/youtrack/issue/LL-636

### CODE REVIEW

File: worker/src/handlers/orgUsageTracker/orgUsageTracker.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+          { $out: 'statementOrgSamples' }
+        ])
+        .exec();
+      const stats = await StatementOrgSample.collection.stats();
+
+      orgStats.avgObjSizes.push(stats.avgObjSize || 0);
+
+      // HELP: @asahd Please add comment about this accepting logic
+      const orgSampleCount = sampleStatementCounts[orgStats.organisation] || 0;
+      const orgSamplePercentage = 100 * orgSampleCount / SAMPLE_SIZE;
```
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply worker/src/handlers/orgUsageTracker/orgUsageTracker.js]: Exactly!

File: worker/src/handlers/orgUsageTracker/orgUsageTracker.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+      if (orgSamplePercentage >= orgStats.totalPercentage - 2) {
```
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply worker/src/handlers/orgUsageTracker/orgUsageTracker.js]: OK

File: worker/src/handlers/orgUsageTracker/orgUsageTracker.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+      await runSample(orgStatsList);
+      totalRunSamples += 1;
+    }
+    logger.info(`Ran ${totalRunSamples} samples`);
+
+    const withTotalUsage = totalUsage(orgStatsList);
+
+    // Save usage stats data
+    await saveOrganisationsWithUsageStats(organisations, withTotalUsage);
+    await OrgUsageStats.create(withTotalUsage);
```
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply worker/src/handlers/orgUsageTracker/orgUsageTracker.js]: OK. Don't we need history of usage stats?

File: worker/src/handlers/orgUsageTracker/totalUsage.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+const toDict = orgStatsList =>
+  orgStatsList.reduce((acc, orgStats) => ({
+    ...acc,
+    [orgStats.organisation]: {
+      ...orgStats,
+      totalCount: 0,
```
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply worker/src/handlers/orgUsageTracker/totalUsage.js]: Yes, this is not necessary and is duplicated.
The reason why this function initialises them again is to make the default function of this file idempotent.

File: worker/src/handlers/orgUsageTracker/orgUsageTracker.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply worker/src/handlers/orgUsageTracker/orgUsageTracker.js]: @ht2 said that history is required

File: worker/src/handlers/orgUsageTracker/initializeOrgStatsList.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ * @async
+ * @param {lib/models/Organisation[]} organisations
+ * @returns {Promise<Object[]>} initialized orgStats list
+ */
+export default async (organisations) => {
+  const isoDate = (new Date()).toISOString();
+
+  const aggregated = await Statement
+    .aggregate([
+      { $group: { _id: '$organisation', count: { $sum: 1 } } }
```
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply worker/src/handlers/orgUsageTracker/initializeOrgStatsList.js]: Use `find({organisation: org._id}, {...}).count()` in organisations.reduce

**Comment by h-kanazawa** [comment]
> I also want to ask why you decided to use a queue job instead of just running the logic from the scheduler.

Scheduler instance should do only scheduling.

File: worker/src/handlers/orgUsageTracker/totalUsage.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * @param {Object[]} orgStatsList - orgStats list
+ * @returns {Object} organisation tree
+ */
+const toTree = (orgStatsList) => {
+  const dict = toDict(orgStatsList);
+  const list = Object.values(dict);
+  list.forEach((orgStat) => {
+    if (orgStat.parentOrganisation) {
+      dict[orgStat.parentOrganisation].children.push(orgStat);
```
[the remark this answers is not in the document]
Comment (h-kanazawa) [reply worker/src/handlers/orgUsageTracker/totalUsage.js]: This tree and the processing order consider grand children/parents.

File: worker/src/handlers/orgUsageTracker/initializeOrgStatsList.js [hunk 7]
```diff
@@ -8,27 +8,24 @@ import { Statement } from 'lib/models';
...
+  const totalStatements = await Statement.count();
 
-  const statementCounts = aggregated.reduce((acc, r) => ({ ...acc, [r._id.toString()]: r.count }), {});
-  const totalStatements = aggregated.reduce((acc, r) => acc + r.count, 0);
+  const orgStatsList = await organisations.reduce(async (accP, org) => {
+    const acc = await accP;
+
+    const ownCount = await Statement
+      .find({ organisation: org._id }, { _id: 0, organisation: 1 })
+      .hint({ organisation: 1, timestamp: -1, _id: 1 })
```
Comment (ryansmith94) [inline worker/src/handlers/orgUsageTracker/initializeOrgStatsList.js]: For James & Asahd's benefit. I've confirmed in our docs that this index should exist.

**Comment by ryansmith94** [comment]
FYI @Ian247 this requires https://ht2labs.myjetbrains.com/youtrack/issue/LL-706

**Comment by h-kanazawa** [comment]
@Ian247 @ht2 I fixed LL-706 at https://github.com/LearningLocker/learninglocker/pull/1295/commits/264b5fa792d56f12d3272ce4bc8f35208fb3ba08

**Comment by h-kanazawa** [comment]
> Can we move **this** outside of the callback so that both scheduled tasks just use a single redis client created once

@ht2 do you mean the line `const redisClient = redis.createClient();` at "this"?

File: worker/src/handlers/orgUsageTracker/initializeOrgStatsList.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+export default async (organisations) => {
+  const runDate = (new Date()).toISOString();
+
+  const totalStatements = await Statement.count();
+
+  const orgStatsList = await organisations.reduce(async (accP, org) => {
+    const acc = await accP;
+
+    const ownCount = await Statement
+      .find({ organisation: org._id }, { _id: 0, organisation: 1 })
```
Comment (ht2) [inline worker/src/handlers/orgUsageTracker/initializeOrgStatsList.js]:
```suggestion
      .find({ organisation: org._id }, { _id: 0, organisation: 1 })
      .read('secondaryPreferred')
```

**Pull request merged**